### ISSUE TITLE
Not able to use the correlation1d-tf

### ISSUE THREAD

**Issue by essaanas1** [luoru/correlation1d-tf#3]
Hi zhongfei, 
I am trying to use the correlation1d layer that you provided in your tensorflow source but I am not able to implement the DispNetC or FlowNetC correlation feature correctly. I am using the the following code for the correlation layer:

```
def Flow_Net(left_image_batch, right_image_batch):
    with tf.variable_scope("conv1") as scope:
        conv1a = conv2d(left_image_batch, [7, 7, 3, 64], strides=2)
        scope.reuse_variables()
        conv1b = conv2d(right_image_batch, [7, 7, 3, 64], strides=2)
    with tf.variable_scope("conv2") as scope:
        conv2a = conv2d(conv1a, [5, 5, 64, 128], strides=2)
        scope.reuse_variables()
        conv2b = conv2d(conv1b, [5, 5, 64, 128], strides=2)   
    with tf.variable_scope("conv_redir"):
        conv_redir = conv2d(conv2a, [1, 1, 128, 64], strides=1)
    with tf.name_scope("corr"):
        corr = tf.correlation1d(conv2a, conv2b, 1, 21,1,1,21)
    with tf.variable_scope("Corr_C"):
        corr1 = conv2d(corr, [1, 1, 41, 64], strides=1)
..........
```

With this configuration the network will be compiled and I can run it but when it runs I got the correlation feature padded with 21 pixels. What I could do to have the feature represented without padding. Note that if I remove the pad the resulting correlation feature will be of size (bz, 96, 150, 41) which is basically the clip version of the feature. of (bz, 96, 192, 41).

I appreciate your feedback.

Essa

**Comment by luoru** [comment]
The tensor format should be N, H, W, C, not N, C, H, W

**Comment by essaanas1** [comment]
Hi zhongfei,

Yes, my feature tensor is already N,H,W,C. Both conv2a, conv2b in
`corr = tf.correlation1d(conv2a, conv2b, 1, 21,1,1,21)`
are (bz, 96,192,128) where 128 is the channels (C)
Could you please explain what you mean?

Your feedback is most appreciated.
Essa

**Comment by luoru** [comment]
if you want to keep tensor size (h * w) with the stride 1 the pad size shoud be same with max_disp, the output channels would be 2 * max_disp + 1, if your max_disp is 20, just use as:
corr =  tf.correlation1d(conv2a, conv2b, 1,  20, 1, 1, 20)
conv2a: [b, h, w, c]
corr: [b, h, w, 2*max_disp + 1]

an optional tf version:
def corr1d_tf(feature1, feature2, d=4):
    b, h, w, c = tf.unstack(tf.shape(feature1))
    feature2 = tf.pad(feature2, [[0, 0], [0, 0], [d, d], [0, 0]], 'CONSTANT')
    cv = []
    for i in range(2 * d + 1):
        cv.append(tf.reduce_mean(feature1 * feature2[:, :, 2 * d - i: 2 * d - i + w, :], axis=-1, keepdims=True))
    return tf.concat(cv, axis=-1)

**Comment by essaanas1** [comment]
Hi,
I am following the instruction that you mentioned as 

```
corr = tf.correlation1d(conv2a, conv2b, 1, 20, 1, 1, 20)
conv2a: [b, h, w, c]
corr: [b, h, w, 2*max_disp + 1]
```
but I am getting the correlation features sifted by 20 pixels, I will send you an image of the generated feature on your email because the site is not allowing to upload images. The second code that you are suggesting using for loop is not applicable for my case because I want to run the code in a model and using loop is not an option. 
My question is: are you saying that using the code: 
`corr = tf.correlation1d(conv2a, conv2b, 1, 20, 1, 1, 20)`
will not produce a shifted correlation feature?!

Thanks in advance for any feedback regarding the issue

Essa

**Comment by essaanas1** [comment]
Hi,

attached the correlation feature generated.

Regard,

Essa
